### ISSUE TITLE
choice_loader sets form model property to label instead of choice

### ISSUE THREAD

**Issue by tistre** [symfony/symfony#17847]
When I use the `ChoiceType` with the `choices` setting in a form, my form data object’s property is correctly set to the choice (i.e. the option value, not the option label). But with `choice_loader`, the property is set to the option label instead.

The example below contains two HTML `select` elements, one using `choices` and the other one using `choice_loader`. If you select “Yes” in both fields and submit the form, the data object (dumped in the template) looks like this:

```php
AppBundle\Entity\Task Object
(
    [field1] => Yes
    [field2] => y
)
```

I think this is a bug, I expect this output instead:

```php
AppBundle\Entity\Task Object
(
    [field1] => y
    [field2] => y
)
```

I’m running Symfony 3.0.2 on PHP 5.5.29. (Sorry for not providing a unit test, I’m a Symfony newbie.)

Here’s my files. The controller:

```php
<?php
// src/AppBundle/Controller/DefaultController.php

namespace AppBundle\Controller;

use AppBundle\Entity\Task;
use AppBundle\ChoiceLoader;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Symfony\Bundle\FrameworkBundle\Controller\Controller;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Form\Extension\Core\Type\SubmitType;
use Symfony\Component\Form\Extension\Core\Type\ChoiceType;

class DefaultController extends Controller
{
    /**
     * @Route("/", name="homepage")
     */
    public function newAction(Request $request)
    {
        $task = new Task();

        $form = $this->createFormBuilder($task)
            ->add
            (
                'field1', 
                ChoiceType::class, 
                [ 
                    'required' => true,
                    'choice_loader' => new ChoiceLoader()
                ]
            )
            ->add
            (
                'field2', 
                ChoiceType::class, 
                [ 
                    'required' => true,
                    'choices'  =>
                    [
                        'Maybe' => 'm',
                        'Yes' => 'y',
                        'No' => 'n',
                    ]
                ]
            )
            ->add('save', SubmitType::class, [ 'label' => 'Submit' ])
            ->getForm();

        $form->handleRequest($request);

        $msg = 'not submitted';

        if ($form->isSubmitted() && $form->isValid()) 
        {
            $msg = print_r($task, true);
        }

        return $this->render('default/new.html.twig', 
        [
            'form' => $form->createView(),
            'msg' => $msg
        ]);
    }
}
```

The form data class:

```php
<?php
// src/AppBundle/Entity/Task.php

namespace AppBundle\Entity;

class Task
{
    public $field1;
    public $field2;
}
```

The choice loader:

```php
<?php
// src/AppBundle/ChoiceLoader.php

namespace AppBundle;

use Symfony\Component\Form\ChoiceList\ArrayChoiceList;
use Symfony\Component\Form\ChoiceList\Loader\ChoiceLoaderInterface;


class ChoiceLoader implements ChoiceLoaderInterface
{
    protected $choices =
    [
        'Maybe' => 'm',
        'Yes' => 'y',
        'No' => 'n'
    ];

    public function loadChoiceList($value = null)
    {
        return new ArrayChoiceList($this->choices);
    }

    public function loadChoicesForValues(array $values, $value = null)
    {
        $result = [ ];
        
        foreach ($values as $val)
        {
            $key = array_search($val, $this->choices, true);
            
            if ($key !== false)
                $result[ ] = $key;
        }

        return $result;
    }

    public function loadValuesForChoices(array $choices, $value = null)
    {
        $result = [ ];
        
        foreach ($choices as $label)
        {
            if (isset($this->choices[ $label ]))
                $result[ ] = $this->choices[ $label ];
        }

        return $result;
    }
}
```

And finally, the template:

```twig
{# app/Resources/views/default/new.html.twig #}

{{ form_start(form) }}
{{ form_widget(form) }}
{{ form_end(form) }}

<pre>{{ msg }}</pre>
```

**Comment by HeahDude** [comment]
The problem come from you loaded choices :
```php
public function loadChoicesForValues(array $values, $value = null)
    {
        $result = [ ];

        foreach ($values as $val)
        {
            $key = array_search($val, $this->choices, true);

            if ($key !== false)
                $result[ ] = $key; // this is wrong, you return the label
                // use instead :
                $result[] = $this->choices[$key];
        }

        return $result;
    }

**Comment by HeahDude** [comment]
Ok I misread sorry

**Comment by HeahDude** [comment]
An array choice list is like `array($value => $choice)`

**Comment by tistre** [comment]
Thanks so much for the explanation – and sorry for me reporting this as a bug. I just got the `ChoiceLoaderInterface` wrong.

My real use case is more sophisticated, by the way, that’s why I’m going with `choice_loader`. (I’m populating choices from a large database table and don’t want to load all choices for validation.)

For reference, this issue helped me understand what’s meant by “choice”, “value” and “label”:

[Fix ChoiceType terminology #6144](https://github.com/symfony/symfony-docs/issues/6144)

Everything works fine with this code:

```php
<?php
// src/AppBundle/ChoiceLoader.php

namespace AppBundle;

use Symfony\Component\Form\ChoiceList\ArrayChoiceList;
use Symfony\Component\Form\ChoiceList\Loader\ChoiceLoaderInterface;


class ChoiceLoader implements ChoiceLoaderInterface
{
    protected $choices =
    [
        'Maybe' => 'm',
        'Yes' => 'y',
        'No' => 'n',
    ];

    public function loadChoiceList($value = null)
    {
        return new ArrayChoiceList($this->choices);
    }

    public function loadChoicesForValues(array $values, $value = null)
    {
        $result = [ ];
        
        foreach ($values as $val)
        {
            $key = array_search($val, $this->choices, true);
            
            if ($key !== false)
                $result[ ] = $val;
        }

        return $result;
    }

    public function loadValuesForChoices(array $choices, $value = null)
    {
        $result = [ ];
        
        foreach ($choices as $choice)
        {
            $key = array_search($choice, $this->choices, true);
            
            if ($key !== false)
                $result[ ] = $choice;
        }

        return $result;
    }
}
```

**Comment by tistre** [comment]
For reference, here’s my blog post containing a fully-fledged and working `choice_loader` example:

[A dropdown for large lists in a Symfony 3 form with choice_loader and Select2](http://www.strehle.de/tim/weblog/archives/2016/02/24/1588)
